### ISSUE TITLE
All Gather test hangs after many runs in loop

### ISSUE THREAD

**Issue by SeanNijjar** [tenstorrent/tt-metal#6388]
This hang was discovered when trying a potential workaround (disabling program caching) for #6363. 

`snijjar/issue-6363`

Disable program caching

The following config hangs reliably after about 120 loop iterations

tests/tt_eager/python_api_testing/unit_testing/misc/test_all_gather.py::test_all_gather_on_t3000_post_commit_looping[500-mem_config1-input_dtype0-8-1-input_shape2-3-layout2] (you may need to manually update the `num_iters` argument parametrization to support 500)

This hang was seen during a local run of the post-commit test cases:
`pytest tests/tt_eager/python_api_testing/unit_testing/misc/test_all_gather.py -k "post_commit"`

It may be reproducible with the standalone test command listed first.

## Enabling watcher avoids the hang. 

Note on priority. There are three concerns with this:
- This isn't _strictly_ a blocker for merging this development branch into maiin as I can artificially reduce the number of loop iterations we run for the all-gather test
- This has a reasonable probability of hitting in real workloads given back to back commands being run in sequence
- There is a "workaround" but I don't think it's one we can reliably apply. It doesn't hang with watcher enabled but we'll then want to enable watcher on all models until this is fixed (or maybe just conditionally on hang), hypothetically.

### UPDATE: 
Bidirectional all-gather support was temporarily disabled on main. For testing with fast-dispatch 2, please re-enable this feature. You can revert the line change in `tt_eager/tt_dnn/op_library/all_gather/all_gather_op.hpp` from commit `b989c269a64c6d40a7523a94f3c4ba16b1eefd20` or ask me for assistance, before retesting.

**Comment by SeanNijjar** [comment]
FYI @jliangTT @davorchap 

It hasn't been conclusively ruled runtime (dispatch path), but we've been seeing a lot of issues like this for that part of the stack after having workers and erisc in the same program.

**Comment by jliangTT** [comment]
hey @SeanNijjar. it is indeed concerning but it is also good that we know we are finding bugs. I notice you said you haven't conclusively ruled runtime. Are you still doing more experiments? Or we should get the runtime team to look at this? 

@davorchap , who from the runtime team can take a look at this? Do we need help from other teams?

**Comment by SeanNijjar** [comment]
I don't have any immediate experiments to run to point one way or another besides direct debug. I'm suspecting runtime based on the behaviour and surrounding issues:


- All gather runs successful many times in a loop then hangs deterministically
  - It would be unlikely for it to be an op issue when the hang happens so deterministically after many runs. IMO this points away from op issue. If it doesn't hang in the first call deterministically, or non-deterministically after the first one, op issue seems unlikely.
- This behaviour is similar to other hangs found with dispatcher or other parts of the dispatch pipeline

Quick exchange with @davorchap offline: Since I'm only hitting this when looping the test, and only for one case so far, I'll consider this a non-blocking issue and I'll just adjust my tests to avoid running this many loop iterations for the problem case.

**Comment by jliangTT** [comment]
Ok i am triaging to Eyon as it is unclear its landing spot. might need to triage. Disabling cacheing changes thing makes me think it is a timing issue. So I think a lot of this migth be related to stuff going on in #6360.

**Comment by eyonland** [comment]
I'm going to own this for now.   Let met see if I can dig up a bit more on this.

**Comment by SeanNijjar** [comment]
yes, I'm definitely leaning more towards this being some lower level dispatcher related bug. It possibly might even end up being a duplicate of: https://github.com/tenstorrent-metal/tt-metal/issues/6404

**Comment by jliangTT** [comment]
do we want to retry this now that @aliuTT has root caused this?

**Comment by eyonland** [comment]
@SeanNijjar ,
Did you already run valgrind and callgrind?  If not, would you be able to do that and attach the results?

**Comment by eyonland** [comment]
@tt-aho , can you comment on this?  Why would enabling watcher impact us here?

**Comment by jliangTT** [comment]
i think before we take any step, @SeanNijjar, @aliuTT's fix landed in main. Do you mind see if you can reproduce?

**Comment by TT-BrianLiu** [comment]
Seems like all gather test is hanging quite frequently on the multi-chip CI: https://github.com/tenstorrent-metal/tt-metal/actions/workflows/multi-device-build-and-unit-tests.yaml

Examples:
- https://github.com/tenstorrent-metal/tt-metal/actions/runs/8350422784/job/22856773992

**Comment by SeanNijjar** [comment]
It looks like something recently changed to cause this issue to be more likely to be exposed, based on this pipeline history @TT-BrianLiu shared. It was looking pretty green and then hit a block of failures.

@eyonland, I haven't run valgrind or callgrind against this test. I'm curious what info you're looking to get from those tools? I'm suspecting this is an issue somewhere in the dispatch data-path. 

Given the (seemingly) increased probability of hang on main pipelines, it may make sense to bump up priority. Additionally, maybe I'll have to remove the looping tests as well. Unfortunately, that won't actually solve the issue :/ .

**Comment by TT-BrianLiu** [comment]
@jliangTT Can we escalate this to P0? Any hangs we are seeing with models can't be debugged unless all_gather is stable.

**Comment by SeanNijjar** [comment]
Going to reassign to @aliuTT based on discussion in bugs meeting.

**Comment by SeanNijjar** [comment]
Possibly related to https://github.com/tenstorrent-metal/tt-metal/issues/6472

**Comment by jliangTT** [comment]
Let's upgrade this to P0 - as this is become a key blocker

**Comment by SeanNijjar** [comment]
Just figured I'd post that this is still reproducible on main:

```

2024-03-23 17:21:03.091 | INFO     | tests.tt_eager.python_api_testing.unit_testing.misc.test_all_gather:run_all_gather_on_t3000_impl:121 - Done iteration 378
2024-03-23 17:21:03.092 | INFO     | tests.tt_eager.python_api_testing.unit_testing.misc.test_all_gather:run_all_gather_on_t3000_impl:121 - Done iteration 379
2024-03-23 17:21:03.093 | INFO     | tests.tt_eager.python_api_testing.unit_testing.misc.test_all_gather:run_all_gather_on_t3000_impl:121 - Done iteration 380
2024-03-23 17:21:03.094 | INFO     | tests.tt_eager.python_api_testing.unit_testing.misc.test_all_gather:run_all_gather_on_t3000_impl:121 - Done iteration 381
2024-03-23 17:21:03.095 | INFO     | tests.tt_eager.python_api_testing.unit_testing.misc.test_all_gather:run_all_gather_on_t3000_impl:121 - Done iteration 382
2024-03-23 17:21:03.096 | INFO     | tests.tt_eager.python_api_testing.unit_testing.misc.test_all_gather:run_all_gather_on_t3000_impl:121 - Done iteration 383
2024-03-23 17:21:03.097 | INFO     | tests.tt_eager.python_api_testing.unit_testing.misc.test_all_gather:run_all_gather_on_t3000_impl:121 - Done iteration 384
2024-03-23 17:21:03.098 | INFO     | tests.tt_eager.python_api_testing.unit_testing.misc.test_all_gather:run_all_gather_on_t3000_impl:121 - Done iteration 385
2024-03-23 17:21:03.099 | INFO     | tests.tt_eager.python_api_testing.unit_testing.misc.test_all_gather:run_all_gather_on_t3000_impl:121 - Done iteration 386
^C^Z
[1]+  Stopped                 pytest tests/tt_eager/python_api_testing/unit_testing/misc/test_all_gather.py -k 'post_commit and looping and 500-mem_config1-input_dtype0-8-1-input_shape2-3-layout2'
```

**Comment by davorchap** [comment]
> Just figured I'd post that this is still reproducible on main:
> 
> 
> 
> ```
> 
> 
> 
> 2024-03-23 17:21:03.091 | INFO     | tests.tt_eager.python_api_testing.unit_testing.misc.test_all_gather:run_all_gather_on_t3000_impl:121 - Done iteration 378
> 
> 2024-03-23 17:21:03.092 | INFO     | tests.tt_eager.python_api_testing.unit_testing.misc.test_all_gather:run_all_gather_on_t3000_impl:121 - Done iteration 379
> 
> 2024-03-23 17:21:03.093 | INFO     | tests.tt_eager.python_api_testing.unit_testing.misc.test_all_gather:run_all_gather_on_t3000_impl:121 - Done iteration 380
> 
> 2024-03-23 17:21:03.094 | INFO     | tests.tt_eager.python_api_testing.unit_testing.misc.test_all_gather:run_all_gather_on_t3000_impl:121 - Done iteration 381
> 
> 2024-03-23 17:21:03.095 | INFO     | tests.tt_eager.python_api_testing.unit_testing.misc.test_all_gather:run_all_gather_on_t3000_impl:121 - Done iteration 382
> 
> 2024-03-23 17:21:03.096 | INFO     | tests.tt_eager.python_api_testing.unit_testing.misc.test_all_gather:run_all_gather_on_t3000_impl:121 - Done iteration 383
> 
> 2024-03-23 17:21:03.097 | INFO     | tests.tt_eager.python_api_testing.unit_testing.misc.test_all_gather:run_all_gather_on_t3000_impl:121 - Done iteration 384
> 
> 2024-03-23 17:21:03.098 | INFO     | tests.tt_eager.python_api_testing.unit_testing.misc.test_all_gather:run_all_gather_on_t3000_impl:121 - Done iteration 385
> 
> 2024-03-23 17:21:03.099 | INFO     | tests.tt_eager.python_api_testing.unit_testing.misc.test_all_gather:run_all_gather_on_t3000_impl:121 - Done iteration 386
> 
> ^C^Z
> 
> [1]+  Stopped                 pytest tests/tt_eager/python_api_testing/unit_testing/misc/test_all_gather.py -k 'post_commit and looping and 500-mem_config1-input_dtype0-8-1-input_shape2-3-layout2'
> 
> ```

Thanks. Is this a P0 ?

**Comment by aliuTT** [comment]
I'll take a look at this. today

**Comment by SeanNijjar** [comment]
Small update here. Even after this gets triaged, I think it will still be a P0. It looks like it's affecting a branch of mine (about to open a PR) that has sharded allgather optimizations. This branch is needed because current sharded all-gather support is purely functional/extremely non-performant.

**Comment by aliuTT** [comment]
I see that when it hangs, it's aways an event not coming back from R chip. Chip varies between different hangs. I was able to dump the remote dispatch and it looks like workers + eth cores are done, so likely something buggy with dispatch -> pull and relay -> tunnelling flow. So it's likely that this is a FD1.3, not sure the value in root causing and getting a fix since we're supposed to get FD2 T3000 parity into main in a week or so. @davorchap can you comment on priority here?

**Comment by SeanNijjar** [comment]
Since we got the okay from @uaydonat to revert, given it would impact perf, I say we go ahead and revert the changes for now and wait for fast dispatch 2 but then make sure to run the allgather tests with 1000+ loop iterations on fast dispatch 2.

**Comment by SeanNijjar** [comment]
Sorry for ping-ponging here. I'm bringing up new all-gather functionality and when I run the program, the all-gather is hanging but when I put prints at the start of all my kernels, and also run with watcher, I see that there are no prints from the kernel. I also took a look at the watcher log and I'm only seeing dispatcher related kernels:

```
k_id[0]: blank
k_id[6]: tt_metal/impl/dispatch/kernels/command_queue_bidirectional_tunneller.cpp
k_id[7]: tt_metal/impl/dispatch/kernels/cq_prefetcher.cpp
k_id[8]: tt_metal/impl/dispatch/kernels/cq_dispatcher.cpp
```

So it looks like the kernels aren't actually being launched.

According to the run log it looks like the program was enqueued:

```
[38;2;000;128;000m               Dispatch[0m | [1m[38;2;128;128;128mTRACE   [0m | [38;2;000;128;000mtt_metal/impl/dispatch/command_queue.cpp:1719[0m - CQ0 received EnqueueCommandType::ALLOCATE_BUFFER in CommandQueueMode::PASSTHROUGH mode
[38;2;000;128;000m               Dispatch[0m | [1m[38;2;128;128;128mTRACE   [0m | [38;2;000;128;000mtt_metal/impl/dispatch/command_queue.cpp:1741[0m - CQ0 running EnqueueCommandType::ALLOCATE_BUFFER
[38;2;000;128;000m               Dispatch[0m | [1m[38;2;128;128;128mTRACE   [0m | [38;2;000;128;000mtt_metal/impl/dispatch/command_queue.cpp:1806[0m - CQ0 running EnqueueCommandType::ALLOCATE_BUFFER complete
[38;2;000;128;000m               Dispatch[0m | [1m[38;2;128;128;128mDEBUG   [0m | EnqueueWriteBuffer for channel 0
[38;2;000;128;000m               Dispatch[0m | [1m[38;2;128;128;128mDEBUG   [0m | EnqueueProgram for channel 0
[38;2;000;128;000m               Dispatch[0m | [1m[38;2;128;128;128mTRACE   [0m | [38;2;000;128;000mtt_metal/impl/dispatch/command_queue.cpp:1806[0m - CQ0 running EnqueueCommandType::ENQUEUE_PROGRAM complete
[38;2;000;128;000m               Dispatch[0m | [1m[38;2;128;128;128mTRACE   [0m | [38;2;000;128;000mtt_metal/impl/dispatch/command_queue.cpp:1719[0m - CQ0 received EnqueueCommandType::FINISH in CommandQueueMode::PASSTHROUGH mode
[38;2;000;128;000m               Dispatch[0m | [1m[38;2;128;128;128mTRACE   [0m | [38;2;000;128;000mtt_metal/impl/dispatch/command_queue.cpp:1741[0m - CQ0 running EnqueueCommandType::FINISH
[38;2;000;128;000m               Dispatch[0m | [1m[38;2;128;128;128mDEBUG   [0m | Finish for command queue 0
```

Do you think this could be related to what you found in your investigation or does it sound like something else?

**Comment by jliangTT** [comment]
Given the current progress in FD2.0, we can target small test such as this against the development branch main (ETA: next monday). (@SeanNijjar @davorchap)

**Comment by jliangTT** [comment]
#6472

**Comment by jliangTT** [comment]
Next step: 

- testing Sean's change against the latest FD2.0 development branch
- ETA: 04/11 (given the current development process, the branch is under debug)

**Comment by aliuTT** [comment]
Currently these tests are all disabled on main, even before FD2 merge. I'll rerun all of them and try to re-enable passing tests for pipelines. Hopefully, they all pass now with FD2...

**Comment by SeanNijjar** [comment]
I've tried rerunning the allgather tests with bidirectional support enabled and unfortunately I'm hitting a non-deterministic hang. 

The hang shows up when running the looping post_commit tests. I don't think the specific config matters but I can reproduce the hang with 

`pytest tests/tt_eager/python_api_testing/unit_testing/misc/test_all_gather.py -k "post_commit and not sharded and looping and False-10000-mem_config1-input_dtype1-4-2-input_shape3-3-layout3"`

usually in a run (but sometimes needs a couple).

Unfortunately, the hang is not reproducible with watcher enabled, even with a minimal configuration of it, even after millions of iterations:
`TT_METAL_WATCHER_DISABLE_ASSERT=1 TT_METAL_WATCHER_DISABLE_DEBUG_STATUS=1 TT_METAL_WATCHER_DISABLE_NOC_SANITIZE=1 pytest tests/tt_eager/python_api_testing/unit_testing/misc/test_all_gather.py -k "post_commit and not sharded and looping and False-1000000-mem_config1-input_dtype1-4-2-input_shape3-3-layout3"`

I'll need to think some more about how to get some more visibility into this.

**Comment by jliangTT** [comment]
this is really interesting that even the minimal config of watcher will perturb the test enough to result in draft difference in behavior. @pgkeller , does this in itself tell us anything? 

also, @SeanNijjar , you mentioned that specific config doesn't matter. Does it mean any config would fail?

**Comment by SeanNijjar** [comment]
By config I mean any all-gather and argument set (shape, concat dim, chip count, link count, etc.), it seems to be roughly generally reproducible without watcher.

In terms of added debug infra, I was thinking that maybe we could make use of the various busy loops in `dataflow_api.h` and have them do some sort of signalling. E.g. if they busy wait more than x cycles, set some register or special value that host can query (on loop exit we can clear it). With large enough threshold, this should realistically have little/no overhead (save for cases where the extra code may affect instruction caching).

**Comment by pgkeller** [comment]
> By config I mean any all-gather and argument set (shape, concat dim, chip count, link count, etc.), it seems to be roughly generally reproducible without watcher.
> 
> In terms of added debug infra, I was thinking that maybe we could make use of the various busy loops in `dataflow_api.h` and have them do some sort of signalling. E.g. if they busy wait more than x cycles, set some register or special value that host can query (on loop exit we can clear it). With large enough threshold, this should realistically have little/no overhead (save for cases where the extra code may affect instruction caching).

That's what debug status does, except it always sets it.  Should be pretty minimal, 1 or 2 loads and a store before and after the loop.

**Comment by SeanNijjar** [comment]
Yea this is definitely surprising that I can't get a hang with the minimal watcher configuration.  Maybe the settings didn't apply properly. @pgkeller, is there somewhere I can check to make sure the environment variables were picked up? (in the watcher log, trace log, elsewhere?). 

Would the only other overheads with the minimal watcher config be:
- kernel setup/teardown -> record Go/Done signals?
- any noc transaction/l1 read timing perturbations due to the (very minimal) overhead of reading the per core state at the watcher time interval
?

**Comment by pgkeller** [comment]
> Yea this is definitely surprising that I can't get a hang with the minimal watcher configuration. Maybe the settings didn't apply properly. @pgkeller, is there somewhere I can check to make sure the environment variables were picked up? (in the watcher log, trace log, elsewhere?).
> 
> Would the only other overheads with the minimal watcher config be:
> 
> * kernel setup/teardown -> record Go/Done signals?
> * any noc transaction/l1 read timing perturbations due to the (very minimal) overhead of reading the per core state at the watcher time interval
>   ?

please sync w/ @tt-dma on the disabling logging
the go/done signals are always enabled, watcher just snoops these, so no overhead there

yes, the only thing left should be the polling overhead only once every <n> seconds

**Comment by SeanNijjar** [comment]
I confirmed, based on input from @tt-dma that `NOC_SANITIZE,PAUSE,STATUS` are all disabled. 
However, looking at the source in `status.h` where we define the DEBUG_STATUS macro, we wrap with

```
#if defined (WATCHER_ENABLED) && !defined(WATCHER_DISABLE_STATUS)
```
However, a search for `WATCHER_DISABLE_STATUS` doesn't seem to surface any results in my workspace (besides the above mentioned one), and commenting out everything under that preprocessor condition out results in build issues (indicating that although watcher thinks `WATCHER_DISABLE_STATUS=1`, the kernel build doesn't see this. I think this explains why I can't reproduce even with the disables specified.

Digging in further with help from David.

**Comment by SeanNijjar** [comment]
Manually removing the kernel code for these watcher features and rerunning with 100k iterations multiple times still results in no hang.

**Comment by SeanNijjar** [comment]
okay we've got some progress here!
Thanks to @tt-dma with some pointers.

It turns out with watcher enabled, we effectively disable inlining! This can _potentially_ lead huge code changes (in my case that must be happening).

I can reproduce the hang with watcher if I disable noc sanitize and if I keep inlining enabled:

```
diff --git a/tt_metal/hw/inc/risc_attribs.h b/tt_metal/hw/inc/risc_attribs.h
index cc7fd6550..3d0b306d9 100644
--- a/tt_metal/hw/inc/risc_attribs.h
+++ b/tt_metal/hw/inc/risc_attribs.h
@@ -37,10 +37,10 @@ inline __attribute__((always_inline)) uint64_t tt_l1_load(volatile tt_uint64_t *
     return v.v;
 }
 
-#if not defined(WATCHER_ENABLED)
+// #if not defined(WATCHER_ENABLED)
     #define FORCE_INLINE inline __attribute__((always_inline))
-#else
-    #define FORCE_INLINE
-#endif
+// #else
+//     #define FORCE_INLINE
+// #endif
 
 #endif // _RISC_ATTRIBS_H_
```

`TT_METAL_WATCHER_DISABLE_NOC_SANITIZE=1 pytest tests/tt_eager/python_api_testing/unit_testing/misc/test_all_gather.py -k "post_commit and not sharded and looping and False-10000-mem_config1-input_dtype1-4-2-input_shape3-3-layout3"`

With this I can get a hang. From the watcher log, I see the allgather kernels running on all chips so for next steps I'll investigate the op directly.

**Comment by jliangTT** [comment]
@tt-dma , does watcher need to disable inline for good reason?

**Comment by tt-dma** [comment]
@jliangTT I didn't realize, but it was disabled a while back to help with code size issues: https://github.com/tenstorrent/tt-metal/commit/b72abf9afa7880c9e1ef471bbe5625a905c0a6fc

I'm currently looking into if we can revert this change, or at least most of it to get better parity for watcher vs non-watcher

**Comment by SeanNijjar** [comment]
Even as a starting point, I think just having a toggle like we do for the other watcher options would still be useful.

**Comment by tt-dma** [comment]
If you rebase, the inlining has been fixed for watcher: https://github.com/tenstorrent/tt-metal/pull/8278

**Comment by SeanNijjar** [comment]
This issue was resolved along with the main work in PR https://github.com/tenstorrent/tt-metal/pull/8303 - this issue ended up being a blocker for getting that mainlined anyways.
